File: iOSDemo/iOSDemo/AdDemo/Ad/Native/Feed(SelfRender)/FeedSelfRenderVC.m
//
//  FeedSelfRenderVC.m
//  iOSDemo
//
//  Created by ltz on 2025/1/21.
//

#import "FeedSelfRenderVC.h"

#import <AnyThinkNative/AnyThinkNative.h>

#import <MJRefresh/MJRefresh.h>
#import "SelfRenderView.h"
#import "DemoOfferAdModel.h"
#import "AdCell.h"
#import "CustomCell.h"
#import "AdLoadConfigTool.h"

@interface FeedSelfRenderVC () <ATNativeADDelegate,UITableViewDelegate,UITableViewDataSource>
 
@property (nonatomic, strong) NSMutableArray <DemoOfferAdModel *> * dataSourceArray;
@property (nonatomic, strong) UITableView * feedTableView;

@end

@implementation FeedSelfRenderVC

// Placement ID
#define Feed_Native_SelfRender_PlacementID @"n67eceed5a282d"

// Scene ID, optional, can be generated in the backend. Pass empty string if not available
#define Feed_Native_SelfRender_SceneID @""
 
- (void)dealloc {
   NSLog(@"FeedSelfRenderVC dealloc");
   // Destroy unreleased ads
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.dataSourceArray.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    
    if (self.dataSourceArray[indexPath.row].isNativeAd) {
        // Ad cell, dynamic height
        return UITableViewAutomaticDimension;
    }
    // Other cells, custom height
    return 200;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    DemoOfferAdModel *offerModel = self.dataSourceArray[indexPath.row];
    
    if (offerModel.isNativeAd) {
        // New ad available, can refresh
        ATNativeAdOffer *offer = [self getOfferAndLoadNext];
        if (offer) {
            offerModel.nativeADView = [self getNativeADViewWithOffer:offer];
            offerModel.offer = offer;
        }
        
        AdCell *cell = [[AdCell alloc] initWithStyle:0 reuseIdentifier:@"AdCell"];
        ATDemoLog(@"adView之前--%@",cell.adView);
        cell.adView = offerModel.nativeADView;
        ATDemoLog(@"adView之后--%@",cell.adView);
        return cell;
        
    }else{
        CustomCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CustomCell"];
        return cell;
    }
}
   
/// Get ATNativeADView object through ad offer
/// - Parameter offer: The obtained ad offer
- (ATNativeADView *)getNativeADViewWithOffer:(ATNativeAdOffer *)offer {
   
    // Initialize config configuration
    ATNativeADConfiguration *config = [[ATNativeADConfiguration alloc] init];
    // Pre-layout for native ad
    config.ADFrame = CGRectMake(0, 0, SelfRenderViewWidth, SelfRenderViewHeight);
    // Pre-layout for video player, recommend to layout again after adding to custom view
    config.mediaViewFrame = CGRectMake(0, 0, SelfRenderViewMediaViewWidth, SelfRenderViewMediaViewHeight);
    config.delegate = self;
    config.rootViewController = self;
    // Set auto-play only in WiFi mode, effective for some ad platforms
    config.videoPlayType = ATNativeADConfigVideoPlayOnlyWiFiAutoPlayType;
    
    // [Manual Layout] Precisely set logo size and position, choose one implementation with [Masonry Method] below
//    config.logoViewFrame = CGRectMake(kScreenW-40-10-15, SelfRenderViewHeight-50-10, 40, 40);
    
    // Set ad platform logo position preference (some ad platforms cannot be precisely set, use the code below, Demo examples all show bottom-right corner)
    // Only when logoUrl or logo has value in material offer, can be set through SelfRenderView layout, otherwise use examples in this method for precise control or preference position setting.
    [ATAPI sharedInstance].preferredAdLogoPosition = ATAdLogoPositionBottomRightCorner;
    
    // Set ad identifier coordinates x and y, effective for some ad platforms
    // config.adChoicesViewOrigin = CGPointMake(10, 10);
     
    // Print material components
    NSDictionary *offerInfoDict = [Tools getOfferInfo:offer];
    ATDemoLog(@"🔥🔥🔥--自渲染广告素材，展示前：%@",offerInfoDict);
    
    // Create self-render view and assign values based on offer information
    SelfRenderView *selfRenderView = [[SelfRenderView alloc] initWithOffer:offer];
    
    // Create ad nativeADView
    // Get native ad display container view
    ATNativeADView *nativeADView = [[ATNativeADView alloc] initWithConfiguration:config currentOffer:offer placementID:Feed_Native_SelfRender_PlacementID];
    
    // Create container array for clickable components
    NSMutableArray *clickableViewArray = [NSMutableArray array];
    
    // Get mediaView, need to add to self-render view manually, must call
    UIView *mediaView = [nativeADView getMediaView];
    if (mediaView) {
        // Assign and layout
        selfRenderView.mediaView = mediaView;
    }
    
    // Set UI controls that need to register click events, better not to add the entire parent view of the feed to click events, otherwise clicking the close button may still trigger the feed click event.
    [clickableViewArray addObjectsFromArray:@[selfRenderView.iconImageView,
                                              selfRenderView.titleLabel,
                                              selfRenderView.textLabel,
                                              selfRenderView.ctaLabel,
                                              selfRenderView.mainImageView]];
    
    // Register click events for UI controls
    [nativeADView registerClickableViewArray:clickableViewArray];
    
    // Bind components
    [self prepareWithNativePrepareInfo:selfRenderView nativeADView:nativeADView];
    
    // Render ad
    [offer rendererWithConfiguration:config selfRenderView:selfRenderView nativeADView:nativeADView];
    
    // [Masonry Method] Precisely set logo size and position, choose one implementation with [Manual Layout] above, call after rendering ad, operate based on nativeADView
    [nativeADView.logoImageView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.right.bottom.mas_equalTo(nativeADView).mas_offset(-25);
        make.width.height.mas_equalTo(15);
    }];
 
    // Hide logo, use with caution, please read documentation: Self-render Ad Considerations
    //nativeADView.logoImageView.hidden = YES;
  
    return nativeADView;
}

#pragma mark - Bind Components
- (void)prepareWithNativePrepareInfo:(SelfRenderView *)selfRenderView nativeADView:(ATNativeADView *)nativeADView {
    // Which components need binding and which don't, please refer to documentation: Native Ad Considerations
    ATNativePrepareInfo *info = [ATNativePrepareInfo loadPrepareInfo:^(ATNativePrepareInfo * prepareInfo) {
        prepareInfo.textLabel = selfRenderView.textLabel;
        prepareInfo.advertiserLabel = selfRenderView.advertiserLabel;
        prepareInfo.titleLabel = selfRenderView.titleLabel;
        prepareInfo.ratingLabel = selfRenderView.ratingLabel;
        prepareInfo.iconImageView = selfRenderView.iconImageView;
        prepareInfo.mainImageView = selfRenderView.mainImageView;
        prepareInfo.ctaLabel = selfRenderView.ctaLabel;
        prepareInfo.dislikeButton = selfRenderView.dislikeButton;
        prepareInfo.mediaView = selfRenderView.mediaView;
    }];
    [nativeADView prepareWithNativePrepareInfo:info];
}

#pragma mark - Data Processing

/// Add data source
/// - Parameter isSuccess: Whether ad loading is successful
- (void)setDataWithRequest:(BOOL)isSuccess {
    if (isSuccess) {
        // Ad loading successful, get offer and add data model
        ATNativeAdOffer *offer = [self getOfferAndLoadNext];
        if (offer) {
            DemoOfferAdModel *offerModel = [[DemoOfferAdModel alloc] init];
            offerModel.nativeADView = [self getNativeADViewWithOffer:offer];
            offerModel.offer = offer;
            // Mark as ad
            offerModel.isNativeAd = YES;
            [self.dataSourceArray addObject:offerModel];
        }
    }
    
    // Add non-ad models, simulate developer's own business cells
    for (int i = 0; i < 3; i ++) {
        DemoOfferAdModel *offerModel1 = [[DemoOfferAdModel alloc] init];
        offerModel1.isNativeAd = NO;
        [self.dataSourceArray addObject:offerModel1];
    }
    [self.feedTableView reloadData];
}

/// Get offer and initiate next load
- (ATNativeAdOffer *)getOfferAndLoadNext {
    
    // Scene statistics feature, optional integration
    [[ATAdManager sharedManager] entryNativeScenarioWithPlacementID:Feed_Native_SelfRender_PlacementID scene:Feed_Native_SelfRender_SceneID];
    
    ATNativeAdOffer *offer = [[ATAdManager sharedManager] getNativeAdOfferWithPlacementID:Feed_Native_SelfRender_PlacementID];
    // Load next
    [self loadNativeAd];
    
    return offer;
}

/// User clicks close, remove ad from list
/// - Parameter nativeADView: Ad adView
- (void)removeAd:(ATNativeADView *)nativeADView {
    for (int i=0; i<self.dataSourceArray.count; i++) {
        DemoOfferAdModel *offerModel = self.dataSourceArray[i];
        if (offerModel.isNativeAd && offerModel.nativeADView == nativeADView) {
            
            if (nativeADView && nativeADView.superview) {
                [nativeADView removeFromSuperview];
            }
            // Destroy ad view
            [nativeADView destroyNative];
            // Destroy offer
            offerModel.offer = nil;
            offerModel.nativeADView = nil;
            
            [self.dataSourceArray removeObject:offerModel];
            [self.feedTableView reloadData];
             
            break;
        }
    }
}
  
/// Print related information for testing
/// - Parameters:
///   - offer: Ad material
///   - nativeADView: Ad object view
- (void)printNativeAdInfoAfterRendererWithOffer:(ATNativeAdOffer *)offer nativeADView:(ATNativeADView *)nativeADView {
    ATNativeAdRenderType nativeAdRenderType = [nativeADView getCurrentNativeAdRenderType];
    if (nativeAdRenderType == ATNativeAdRenderExpress) {
        ATDemoLog(@"⚠️⚠️⚠️--这是原生模板了");
    } else {
        ATDemoLog(@"✅✅✅--这是原生自渲染");
    }
    BOOL isVideoContents = [nativeADView isVideoContents];
    
    // Print all material content
    [Tools printNativeAdOffer:offer];
    ATDemoLog(@"🔥--是否为原生视频广告：%d",isVideoContents);
}

#pragma mark - Placement Delegate Callbacks
/// Placement loading completed
/// - Parameter placementID: Placement ID
- (void)didFinishLoadingADWithPlacementID:(NSString *)placementID {
    BOOL isReady = [[ATAdManager sharedManager] nativeAdReadyForPlacementID:placementID];
    [self showLog:[NSString stringWithFormat:@"didFinishLoadingADWithPlacementID:%@ SelfRender 是否准备好:%@", placementID,isReady ? @"YES":@"NO"]];
    
    if (self.feedTableView.mj_footer.refreshing == YES) {
        [self.feedTableView.mj_footer endRefreshing];
        // Request successful, pass YES to get ad data
        [self setDataWithRequest:YES];
    }
}
 
/// Placement loading failed
/// - Parameters:
///   - placementID: Placement ID
///   - error: Error information
- (void)didFailToLoadADWithPlacementID:(NSString *)placementID error:(NSError *)error {
    ATDemoLog(@"didFailToLoadADWithPlacementID:%@ error:%@", placementID, error);
    [self showLog:[NSString stringWithFormat:@"didFailToLoadADWithPlacementID:%@ errorCode:%ld", placementID, error.code]];
    
    if (self.feedTableView.mj_footer.refreshing == YES) {
        [self.feedTableView.mj_footer endRefreshing];
        // Request failed
        [self setDataWithRequest:NO];
    }
}

/// Received display revenue
/// - Parameters:
///   - placementID: Placement ID
///   - extra: Extra information dictionary
- (void)didRevenueForPlacementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didRevenueForPlacementID:%@ with extra: %@", placementID,extra);
    [self showLog:[NSString stringWithFormat:@"didRevenueForPlacementID:%@", placementID]];
}

#pragma mark - Native Ad Event Callbacks

/// Native ad displayed
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
- (void)didShowNativeAdInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didShowNativeAdInAdView:%@ extra:%@",placementID,extra);
    [self showLog:[NSString stringWithFormat:@"didShowNativeAdInAdView:%@", placementID]];
    
    
}

/// Native ad close button clicked
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
- (void)didTapCloseButtonInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didTapCloseButtonInAdView:%@ extra:%@", placementID, extra);
    [self showLog:[NSString stringWithFormat:@"didTapCloseButtonInAdView:%@", placementID]];
    
    // Remove from list
    [self removeAd:adView];
}

/// Native ad started playing video
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information dictionary
- (void)didStartPlayingVideoInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didStartPlayingVideoInAdView:%@ extra: %@", placementID,extra);
    [self showLog:[NSString stringWithFormat:@"didStartPlayingVideoInAdView:%@", placementID]];
}

/// Native ad video playback ended
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information dictionary
- (void)didEndPlayingVideoInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didEndPlayingVideoInAdView:%@ extra: %@", placementID,extra);
    [self showLog:[NSString stringWithFormat:@"didEndPlayingVideoInAdView:%@", placementID]];
}

/// Native ad user clicked
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information dictionary
- (void)didClickNativeAdInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didClickNativeAdInAdView:%@ extra:%@",placementID,extra);
    [self showLog:[NSString stringWithFormat:@"didClickNativeAdInAdView:%@", placementID]];
}
 
/// Native ad opened or jumped to deep link page
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
///   - success: Whether successful
- (void)didDeepLinkOrJumpInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra result:(BOOL)success {
    ATDemoLog(@"didDeepLinkOrJumpInAdView:placementID:%@ with extra: %@, success:%@", placementID,extra, success ? @"YES" : @"NO");
    [self showLog:[NSString stringWithFormat:@"didDeepLinkOrJumpInAdView:%@, success:%@", placementID, success ? @"YES" : @"NO"]];
}
 
/// Native ad entered full-screen video playback, usually auto-jumps to a playback landing page after clicking video mediaView
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
- (void)didEnterFullScreenVideoInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra{
    ATDemoLog(@"didEnterFullScreenVideoInAdView:%@", placementID);
    [self showLog:[NSString stringWithFormat:@"didEnterFullScreenVideoInAdView:%@", placementID]];
}

/// Native ad exited full-screen video playback
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
- (void)didExitFullScreenVideoInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didExitFullScreenVideoInAdView:%@", placementID);
    [self showLog:[NSString stringWithFormat:@"didExitFullScreenVideoInAdView:%@", placementID]];
}
 
/// Native ad exited detail page
/// - Parameters:
///   - adView: Ad view object
///   - placementID: Placement ID
///   - extra: Extra information
- (void)didCloseDetailInAdView:(ATNativeADView *)adView placementID:(NSString *)placementID extra:(NSDictionary *)extra {
    ATDemoLog(@"didCloseDetailInAdView:%@ extra:%@", placementID, extra);
    [self showLog:[NSString stringWithFormat:@"didCloseDetailInAdView:%@", placementID]];
}

#pragma mark - TableView Data Refresh
- (void)footerRefresh {
    self.feedTableView.mj_footer = [MJRefreshAutoNormalFooter footerWithRefreshingTarget:self refreshingAction:@selector(upFreshLoadMoreData)];
    [self.feedTableView.mj_footer beginRefreshing];
}

/// Load more
- (void)upFreshLoadMoreData {
    [self loadNativeAd];
}
 
/// Load ad
- (void)loadNativeAd {
    NSMutableDictionary * loadConfigDict = [NSMutableDictionary dictionary];
    
    // Set ad request size
    [loadConfigDict setValue:[NSValue valueWithCGSize:CGSizeMake(SelfRenderViewWidth, SelfRenderViewHeight)] forKey:kATExtraInfoNativeAdSizeKey];
    // Request adaptive size native ad (available for some ad platforms) (optional integration)
    [AdLoadConfigTool native_loadExtraConfigAppend_SizeToFit:loadConfigDict];
    
    // KuaiShou native ad swipe and click control
//    [AdLoadConfigTool native_loadExtraConfigAppend_KuaiShou_SlideOrClickAble:loadConfigDict];
    
    [[ATAdManager sharedManager] loadADWithPlacementID:Feed_Native_SelfRender_PlacementID extra:loadConfigDict delegate:self];
}
 
#pragma mark - getter
- (UITableView *)feedTableView {
    if (!_feedTableView) {
        _feedTableView = [[UITableView alloc] init];
        _feedTableView.delegate = self;
        _feedTableView.dataSource = self;
        _feedTableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
        
        // Bind cells
        [_feedTableView registerNib:[UINib nibWithNibName:@"CustomCell" bundle:nil] forCellReuseIdentifier:@"CustomCell"];
        [_feedTableView registerClass:[AdCell class] forCellReuseIdentifier:@"AdCell"];
        
        // Set a reasonable estimated height
        _feedTableView.estimatedRowHeight = SelfRenderViewHeight; // Set a reasonable estimated value
        _feedTableView.rowHeight = UITableViewAutomaticDimension;
    }
    return _feedTableView;
}

- (NSMutableArray *)dataSourceArray {
    if (_dataSourceArray) return _dataSourceArray;
    NSMutableArray *dataSourceArray = [NSMutableArray array];
    return _dataSourceArray = dataSourceArray;
}
 
#pragma mark - Demo UI
- (void)viewDidLoad {
    [super viewDidLoad];
     
    [self.view addSubview:self.feedTableView];
    [self.feedTableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.nbar.mas_bottom);
        make.left.right.bottom.mas_equalTo(self.view);
    }];
    [self footerRefresh];
}
 
@end

File: SwiftDemo/SwiftDemo/AdDemo/Native/FeedSelfRenderVC.swift
//
//  FeedSelfRenderVC.swift
//  SwiftDemo
//
//  Created by ltz on 2025/1/21.
//

import UIKit
import AnyThinkNative
import MJRefresh
import SnapKit

// MARK: - Constants
private let SelfRenderViewWidth: CGFloat = kScreenW
private let SelfRenderViewHeight: CGFloat = 350
private let SelfRenderViewMediaViewWidth: CGFloat = kScreenW
private let SelfRenderViewMediaViewHeight: CGFloat = 175

private let Feed_Native_SelfRender_PlacementID = "n67eceed5a282d"
private let Feed_Native_SelfRender_SceneID = ""

// MARK: - Models
class DemoOfferAdModel: NSObject {
    var nativeADView: ATNativeADView?
    var offer: ATNativeAdOffer?
    var isNativeAd: Bool = false
}

class FeedSelfRenderVC: BaseNormalBarVC {
    
    // MARK: - Properties
    private var dataSourceArray: [DemoOfferAdModel] = []
    
    private lazy var feedTableView: UITableView = {
        let tableView = UITableView()
        tableView.delegate = self
        tableView.dataSource = self
        tableView.separatorStyle = .singleLine
        tableView.register(FeedCustomCell.self, forCellReuseIdentifier: "FeedCustomCell")
        tableView.register(AdCell.self, forCellReuseIdentifier: "AdCell")
        tableView.estimatedRowHeight = SelfRenderViewHeight
        tableView.rowHeight = UITableView.automaticDimension
        return tableView
    }()
    
    deinit {
        ATDemoLog("FeedSelfRenderVC dealloc")
    }
    
    // MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.addSubview(feedTableView)
 
        feedTableView.snp.makeConstraints { make in
            make.top.equalTo(self.nbar!.snp.bottom)
            make.left.right.bottom.equalTo(view)
        }
         
        footerRefresh()
    }
    
    // MARK: - TableView Refresh
    private func footerRefresh() {
        let footer = MJRefreshAutoNormalFooter(refreshingTarget: self, refreshingAction: #selector(upFreshLoadMoreData))
        feedTableView.mj_footer = footer
        feedTableView.mj_footer?.beginRefreshing()
    }
    
    @objc private func upFreshLoadMoreData() {
        loadNativeAd()
    }
    
    // MARK: - Data Handling
    private func setData(isSuccess: Bool) {
        if isSuccess {
            // Load ad successfully, get offer and add data model
            if let offer = getOfferAndLoadNext() {
                let offerModel = DemoOfferAdModel()
                offerModel.nativeADView = getNativeADView(offer: offer)
                offerModel.offer = offer
                offerModel.isNativeAd = true
                dataSourceArray.append(offerModel)
            }
        }
        
        // Add non-ad models, simulate developer's own business cells
        for _ in 0..<3 {
            let offerModel = DemoOfferAdModel()
            offerModel.isNativeAd = false
            dataSourceArray.append(offerModel)
        }
        feedTableView.reloadData()
    }
    
    private func getOfferAndLoadNext() -> ATNativeAdOffer? {
        // Scene statistics, optional integration
        ATAdManager.shared().entryNativeScenario(withPlacementID: Feed_Native_SelfRender_PlacementID, scene: Feed_Native_SelfRender_SceneID)
        
        let offer = ATAdManager.shared().getNativeAdOffer(withPlacementID: Feed_Native_SelfRender_PlacementID)
        // load next
        loadNativeAd()
        
        return offer
    }
    
    private func removeAd(nativeADView: ATNativeADView?) {
        guard let nativeADView = nativeADView else { return }
        
        for (index, offerModel) in dataSourceArray.enumerated() {
            if offerModel.isNativeAd && offerModel.nativeADView == nativeADView {
                
                if nativeADView.superview != nil {
                    nativeADView.removeFromSuperview()
                }
                // Destroy ad view
                nativeADView.destroyNative()
                // Destroy offer
                offerModel.offer = nil
                offerModel.nativeADView = nil
                
                dataSourceArray.remove(at: index)
                feedTableView.reloadData()
                break
            }
        }
    }
    
    // MARK: - Ad Loading and Rendering
    private func loadNativeAd() {
        var loadConfigDict: [String: Any] = [:]
        
        // Set ad request size
        loadConfigDict[kATExtraInfoNativeAdSizeKey] = NSValue(cgSize: CGSize(width: SelfRenderViewWidth, height: SelfRenderViewHeight))
        // Request self-adapting size native ad (available for some ad platforms) (optional integration)
        AdLoadConfigTool.native_loadExtraConfigAppend_SizeToFit(&loadConfigDict)
        
        ATAdManager.shared().loadAD(withPlacementID: Feed_Native_SelfRender_PlacementID, extra: loadConfigDict, delegate: self)
    }
    
    private func getNativeADView(offer: ATNativeAdOffer) -> ATNativeADView {
        // Initialize config configuration
        let config = ATNativeADConfiguration()
        // Pre-layout for native ad
        config.adFrame = CGRect(x: 0, y: 0, width: SelfRenderViewWidth, height: SelfRenderViewHeight)
        // Pre-layout for video player
        config.mediaViewFrame = CGRect(x: 0, y: 0, width: SelfRenderViewMediaViewWidth, height: SelfRenderViewMediaViewHeight)
        config.delegate = self
        config.rootViewController = self
        // Make ad view container fit the ad
        config.sizeToFit = true
        // Set auto-play only in WiFi mode
        config.videoPlayType = .onlyWiFiAutoPlayType
        
        // Set ad platform logo position preference
        ATAPI.sharedInstance().preferredAdLogoPosition = .bottomRightCorner
        
        let offerInfoDict = Tools.getOfferInfo(offer)
        ATDemoLog("🔥🔥🔥--Self-render ad material, before display: \(offerInfoDict)")
        
        // Create self-render view and assign values
        let selfRenderView = SelfRenderView(offer: offer)
        
        // Create ad nativeADView
        let nativeADView = ATNativeADView(configuration: config, currentOffer: offer, placementID: Feed_Native_SelfRender_PlacementID)
        
        // Create clickable views array
        var clickableViewArray: [UIView] = []
        
        // Get mediaView
        if let mediaView = nativeADView.getMediaView() {
            selfRenderView.mediaView = mediaView
        }
        
        // Add clickable views
        clickableViewArray.append(contentsOf: [selfRenderView.iconImageView,
                                               selfRenderView.titleLabel,
                                               selfRenderView.textLabel,
                                               selfRenderView.ctaLabel,
                                               selfRenderView.mainImageView])
        
        // Register click events
        nativeADView.registerClickableViewArray(clickableViewArray)
        
        // Bind components
        prepareWithNativePrepareInfo(selfRenderView: selfRenderView, nativeADView: nativeADView)
        
        // Render ad
        offer.renderer(with: config, selfRenderView: selfRenderView, nativeADView: nativeADView)
        
        // Set logo size and position
        if let logoImageView = nativeADView.logoImageView() {
            logoImageView.snp.remakeConstraints { make in
                make.right.bottom.equalTo(nativeADView).offset(-25)
                make.width.height.equalTo(15)
            }
        }
        
        return nativeADView
    }
    
    private func prepareWithNativePrepareInfo(selfRenderView: SelfRenderView, nativeADView: ATNativeADView) {
        let info = ATNativePrepareInfo.load { prepareInfo in
            prepareInfo.textLabel = selfRenderView.textLabel
            prepareInfo.advertiserLabel = selfRenderView.advertiserLabel
            prepareInfo.titleLabel = selfRenderView.titleLabel
            prepareInfo.ratingLabel = selfRenderView.ratingLabel
            prepareInfo.iconImageView = selfRenderView.iconImageView
            prepareInfo.mainImageView = selfRenderView.mainImageView
            prepareInfo.ctaLabel = selfRenderView.ctaLabel
            prepareInfo.dislikeButton = selfRenderView.dislikeButton
            if let mediaView = selfRenderView.mediaView {
                prepareInfo.mediaView = mediaView
            }
            //Only for Yandex , must bind & render
//            prepareInfo.domainLabel = selfRenderView.domainLabel
//            prepareInfo.warningLabel = selfRenderView.warningLabel
        }
        nativeADView.prepare(with: info)
    }
    
    private func printNativeAdInfoAfterRenderer(offer: ATNativeAdOffer, nativeADView: ATNativeADView) {
        let nativeAdRenderType = nativeADView.getCurrentNativeAdRenderType()
        if nativeAdRenderType == .express {
            ATDemoLog("⚠️⚠️⚠️--Template ad")
        } else {
            ATDemoLog("✅✅✅--Self-render ad")
        }
        let isVideoContents = nativeADView.isVideoContents()
        
        // Print all material content
        Tools.printNativeAdOffer(offer)
        ATDemoLog("🔥--Is native video ad: \(isVideoContents)")
    }
}

// MARK: - UITableViewDelegate, UITableViewDataSource
extension FeedSelfRenderVC: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataSourceArray.count
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let model = dataSourceArray[indexPath.row]
        if model.isNativeAd {
            return UITableView.automaticDimension
        }
        return 156
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let offerModel = dataSourceArray[indexPath.row]
        
        if offerModel.isNativeAd {
            // New ad available, refresh
            if let offer = getOfferAndLoadNext() {
                offerModel.nativeADView = getNativeADView(offer: offer)
                offerModel.offer = offer
            }
            
            let cell = tableView.dequeueReusableCell(withIdentifier: "AdCell", for: indexPath) as! AdCell
            cell.adView = offerModel.nativeADView
            return cell
        } else {
            let cell = tableView.dequeueReusableCell(withIdentifier: "FeedCustomCell", for: indexPath) as! FeedCustomCell
            return cell
        }
    }
}

// MARK: - ATNativeADDelegate
extension FeedSelfRenderVC: ATNativeADDelegate {
    func didFinishLoadingAD(withPlacementID placementID: String) {
        let isReady = ATAdManager.shared().nativeAdReady(forPlacementID: placementID)
        showLog("didFinishLoadingADWithPlacementID:\(placementID) SelfRender isReady:\(isReady ? "YES" : "NO")")
        
        if feedTableView.mj_footer?.isRefreshing == true {
            feedTableView.mj_footer?.endRefreshing()
            setData(isSuccess: true)
        }
    }
    
    func didFailToLoadAD(withPlacementID placementID: String, error: Error) {
        ATDemoLog("didFailToLoadADWithPlacementID:\(placementID) error:\(error)")
        showLog("didFailToLoadADWithPlacementID:\(placementID) errorCode:\((error as NSError).code)")
        
        if feedTableView.mj_footer?.isRefreshing == true {
            feedTableView.mj_footer?.endRefreshing()
            setData(isSuccess: false)
        }
    }
    
    func didRevenue(forPlacementID placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didRevenueForPlacementID:\(placementID) with extra: \(extra)")
        showLog("didRevenueForPlacementID:\(placementID)")
    }
    
    func didShowNativeAd(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didShowNativeAdInAdView:\(placementID) extra:\(extra)")
        showLog("didShowNativeAdInAdView:\(placementID)")
    }
    
    func didTapCloseButton(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didTapCloseButtonInAdView:\(placementID) extra:\(extra)")
        showLog("didTapCloseButtonInAdView:\(placementID)")
        
        removeAd(nativeADView: adView)
    }
    
    func didStartPlayingVideo(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didStartPlayingVideoInAdView:\(placementID) extra: \(extra)")
        showLog("didStartPlayingVideoInAdView:\(placementID)")
    }
    
    func didEndPlayingVideo(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didEndPlayingVideoInAdView:\(placementID) extra: \(extra)")
        showLog("didEndPlayingVideoInAdView:\(placementID)")
    }
    
    func didClickNativeAd(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didClickNativeAdInAdView:\(placementID) extra:\(extra)")
        showLog("didClickNativeAdInAdView:\(placementID)")
    }
    
    func didDeepLinkOrJump(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any], result success: Bool) {
        ATDemoLog("didDeepLinkOrJumpInAdView:placementID:\(placementID) with extra: \(extra), success:\(success ? "YES" : "NO")")
        showLog("didDeepLinkOrJumpInAdView:\(placementID), success:\(success ? "YES" : "NO")")
    }
    
    func didEnterFullScreenVideo(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didEnterFullScreenVideoInAdView:\(placementID)")
        showLog("didEnterFullScreenVideoInAdView:\(placementID)")
    }
    
    func didExitFullScreenVideo(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didExitFullScreenVideoInAdView:\(placementID)")
        showLog("didExitFullScreenVideoInAdView:\(placementID)")
    }
    
    func didCloseDetail(in adView: ATNativeADView, placementID: String, extra: [AnyHashable : Any]) {
        ATDemoLog("didCloseDetailInAdView:\(placementID) extra:\(extra)")
        showLog("didCloseDetailInAdView:\(placementID)")
    }
}

